[TOOLS][TESTS] Add tooling for running flake8 as a pre-commit hook and in CI

elezar opened the pull request:
This PR was a bit of train work from my side, and is a start at getting some basic syntax / error checking into our CI for our Python files. I started with `flake8`, but would like to add `pylint -E`, `black`, and `pyre` at some point.

There have been some errors in recent PRs that increased the time to merge and at least a subset of these would have been detected by these tools.

The PR makes the following specific changes:
* Adds `flake8` as a requirement in the docker image
* Adds `pylint` as a requirement in the docker image
* Adds a collection of scripts to drive these kinds of checks
* adds a `pre-commit-config.yaml` file for enabling checks locally
* Make some change to Python files in `testing/` while checking the commit hooks and scripts.

Note:
* Only modified files are checked. Meaning that we can gradually improve the codebase without having to check in large diffs.

I am also testing a basic TeamCity jobs:

* Intentional failure: https://teamcity.mesosphere.io/viewLog.html?buildId=1102872&buildTypeId=DcosIo_DcosCommons_PrJobs_TestCodeChecksPr&tab=buildLog
* Intentional success: https://teamcity.mesosphere.io/viewLog.html?buildId=1102873&buildTypeId=DcosIo_DcosCommons_PrJobs_TestCodeChecksPr&tab=buildLog
* A template configuration here: https://teamcity.mesosphere.io/viewType.html?buildTypeId=DcosIo_DcosCommons_PrJobs_RunCodeChecks

elezar commented:
@nickbp @benclarkwood @porridge I've added you as reviewers here. Would love to hear your thoughts.

My plan would be to first merge this (with no changes to the CI system) and then work on integrating the checks into the CI.

porridge commented on tools/ci/checks/get_applicable_changes.sh:
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
```
Please do this in Python. It will probably be shorter and easier to see whats going on.

porridge commented on tools/ci/checks/get_base_branch.sh:
```diff
@@ -0,0 +1,21 @@
+current_branch=${CURRENT_GIT_BRANCH:-$( git symbolic-ref --short HEAD )}
+
+base_branch="master"
+real_branch=''
```
Please use double-quote for consistency.

porridge commented on tools/ci/checks/get_base_branch.sh:
```diff
@@ -0,0 +1,21 @@
+    pr_name="${current_branch/pull/pulls}"
+
+    git_repo="${$(git remote get-url origin | sed -e 's/.*://g')//.git/}"
+    output=$( curl "https://api.github.com/repos/${git_repo}/${pr_name}" )
```
Three lines with `$()`, three different styles of spacing inside them :-)

porridge replied:
Note that curl by default reports errors as successes:
```
porridge@beczulka:~$ curl httpstat.us/500;echo;echo $?
500 Internal Server Error
0
porridge@beczulka:~$ 
```

porridge commented on tools/ci/checks/get_base_branch.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+#
```
Remove this line.

porridge commented on tools/ci/checks/get_changeset.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+# Get the set of files changed relative to a specifed git reference.
+# The COMPARE_TO environment variable is used, and if no value is specified
```
Well, the logic is actually slightly different: if HEAD is invalid, then $COMPARE_TO will be ignored, no?

porridge commented on tools/ci/checks/run_check_in_docker.sh:
```diff
@@ -0,0 +1,14 @@
+DOCKER_IMAGE=${DOCKER_IMAGE:-mesosphere/dcos-commons:${DOCKER_TAG}}
+
+
+TOOL=$1
```
Why do we need this?

porridge commented on tools/ci/checks/run_check_in_docker.sh:
```diff
@@ -0,0 +1,14 @@
+    -v $(pwd):/build:ro \
+    -w /build \
+        ${DOCKER_IMAGE} \
+            ${TOOL} $*
```
Why not "$@"?

porridge commented on tools/ci/checks/run_flake8_checks.sh:
```diff
@@ -0,0 +1,6 @@
+
+TOOL_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+${TOOL_DIR}/run_check_in_docker.sh flake8 $*
```
Why not "$@"?

porridge commented on tools/ci/steps/check_python_files.sh:
```diff
@@ -0,0 +1,29 @@
+    echo "${CHANGESET}"
+
+    echo ""
+    echo "Running flake8 on $(echo \"${CHANGESET}\" | wc -w) files:"
```
Again, inconsistent spaces in `$()`

porridge commented on tools/ci/steps/check_python_files.sh:
```diff
@@ -0,0 +1,29 @@
+    echo "Changeset:"
+    echo "${CHANGESET}"
+
+    echo ""
```
Superfluous `""`

elezar commented:
Thanks for the review @porridge. I have addressed most of your concerns -- I think. With regards to:
> This is awesome, although I hope we can agree on a timeline to move to running on all files rather just the changed ones, since changing file A can break file B which do not change but uses it.

Yes, we would definitely want to address these in bulk instead of one file at a time. This PR is more about getting the infrastructure in so that we can do this in a follow-up.

elezar commented:
Since this would have caught the error in #2549, I have gone ahead and enabled `pylint -E` checks here too.

elezar commented on tools/ci/checks/get_base_branch.sh:
```diff
@@ -1,21 +1,45 @@
+        set +x
+    fi
+
+    output=$( curl --silent "https://api.github.com/repos/${GIT_REPO}/${pr_name}" )
```
Will that not pollute the output?

porridge replied:
No. Even without `--silent` curls is well-behaved enough to send diagnostics to stderr:
```
porridge@beczulka:~$ curl --retry 3 --silent httpstat.us/500 | wc ;echo
      0      13     100

porridge@beczulka:~$ curl --retry 3 httpstat.us/500 | wc ;echo
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    25  100    25    0     0     60      0 --:--:-- --:--:-- --:--:--    60
Warning: Transient problem: HTTP error Will retry in 1 seconds. 3 retries 
Warning: left.
100    25  100    25    0     0    121      0 --:--:-- --:--:-- --:--:--   121
Warning: Transient problem: HTTP error Will retry in 2 seconds. 2 retries 
Warning: left.
100    25  100    25    0     0    121      0 --:--:-- --:--:-- --:--:--   121
Warning: Transient problem: HTTP error Will retry in 4 seconds. 1 retries 
Warning: left.
100    25  100    25    0     0    121      0 --:--:-- --:--:-- --:--:--   121
      0      13     100

porridge@beczulka:~$ 
```

mpereira commented on tools/ci/checks/run_flake8_checks.sh:
```diff
@@ -1,6 +1,11 @@
+    -v $(pwd):/build:ro \
+    -w /build \
+        ${DOCKER_IMAGE} \
+            flake8 $@
```
It's a good idea to always quote variables. Maybe we can get ShellCheck running against our bash in the near future too...

elezar replied:
They do behave differently though. In both cases, there is a difference between passing the arguments as a space separated list (`$@`) and a quoted mulit-line


In the case of `pylint`:
```
+ docker run --rm -t -v /Users/elezar/src/mesosphere/dcos-commons:/build:ro -w /build mesosphere/dcos-commons:latest pylint -E testing/sdk_diag.py testing/sdk_hosts.py testing/sdk_install.py testing/sdk_jobs.py testing/sdk_metrics.py testing/sdk_plan.py testing/sdk_security.py testing/sdk_tasks.py testing/sdk_upgrade.py tools/ci/checks/get_applicable_changes.py
```
vs
```
docker run --rm -t -v /Users/elezar/src/mesosphere/dcos-commons:/build:ro -w /build mesosphere/dcos-commons:latest pylint -E 'testing/sdk_diag.py
testing/sdk_hosts.py
testing/sdk_install.py
testing/sdk_jobs.py
testing/sdk_metrics.py
testing/sdk_plan.py
testing/sdk_security.py
testing/sdk_tasks.py
testing/sdk_upgrade.py
tools/ci/checks/get_applicable_changes.py'
```

From http://tldp.org/LDP/abs/html/internalvariables.html
> Same as $*, but each parameter is a quoted string, that is, the parameters are passed on intact, without interpretation or expansion. This means, among other things, that each parameter in the argument list is seen as a separate word.

Which is what I want since each element of `$@` is a separate filename.

porridge replied:
Right, this is the point. Please pass them as individual elements instead. Otherwise we'll be in a world of pain as soon as a file/dir name with space appears.

elezar replied:
I don't quite follow what you mean. What I'm saying is that `$@` already quotes the separate elements and `"$@"` does *not* work for these scripts.

porridge replied:
Here's a test case:
```
porridge@beczulka:~/tmp/x$ mkdir t
porridge@beczulka:~/tmp/x$ touch t/1 't/2 2'
porridge@beczulka:~/tmp/x$ echo 'ls -l "$@"' > quoted
porridge@beczulka:~/tmp/x$ echo 'ls -l $@' > unquoted
porridge@beczulka:~/tmp/x$ bash quoted t/*
-rw-r--r-- 1 porridge porridge 0 lip  2 18:18 t/1
-rw-r--r-- 1 porridge porridge 0 lip  2 18:18 t/2 2
porridge@beczulka:~/tmp/x$ bash unquoted t/*
ls: nie ma dostępu do 't/2': Nie ma takiego pliku ani katalogu
ls: nie ma dostępu do '2': Nie ma takiego pliku ani katalogu
-rw-r--r-- 1 porridge porridge 0 lip  2 18:18 t/1
porridge@beczulka:~/tmp/x$ bash quoted "t/1 t/2 2"
ls: nie ma dostępu do 't/1 t/2 2': Nie ma takiego pliku ani katalogu
porridge@beczulka:~/tmp/x$ bash unquoted "t/1 t/2 2"
ls: nie ma dostępu do 't/2': Nie ma takiego pliku ani katalogu
ls: nie ma dostępu do '2': Nie ma takiego pliku ani katalogu
-rw-r--r-- 1 porridge porridge 0 lip  2 18:18 t/1
porridge@beczulka:~/tmp/x$ 
```

mpereira commented on tools/ci/checks/get_applicable_changes.py:
```diff
@@ -0,0 +1,99 @@
+
+
+def flatten_file_list(file_args: List[str]) -> List[str]:
+
```
What does pep8 say about this? I'd vote for no blank too.

elezar replied:
`pep8` (or `flake8`) doesn't complain as long as there are no internal double empty lines. My plan is to enable [black](https://github.com/ambv/black) at some point so that we don't need discussions about formatting anymore. It also only has 1 setting. Still want to be able to do this "automatically" though to reduce the developer burden.

mpereira commented on tools/ci/checks/get_base_branch.sh:
```diff
@@ -0,0 +1,45 @@
+
+    if [ -z ${GIT_REPO} ]; then
+        set -x
+        git_repo=$( git remote get-url origin )
```
fwiw it is also good practice to quote command substitutions, e.g. `git_repo="$(git remote get-url origin)"`.

elezar commented:
Thanks for the reviews @porridge @mpereira @benclarkwood. I think I have addressed all of your comments now.

mpereira commented:
@porridge it looks like your feedback was addressed so I'm merging this. Let me know if this is not the case!

elezar commented:
> We still need to push mesosphere/dcos-commons:latest with these Dockerfile changes to have flake8 installed for the check script.

The changes are built by docker-hub.
